### DISCUSSION TITLE
Add a profile for notable

### DISCUSSION THREAD

**Discussion by githlp** [netblue30/firejail#4812] [Q&A] [answered]
when i try to start it like this
`firejail --net=none /opt/Notable/notable`
then i get this error
`Check failed: sys_chroot("/proc/self/fdinfo/") == 0`

if i try to build a profile for the app
`firejail --build profile-file /opt/Notable/notable`
then the following error occurs
`/usr/bin/strace: Can't stat 'profile-file': Datei oder Verzeichnis nicht gefunden`
`Error: cannot run the sandbox`

**Comment by rusty-snake** [comment]
>  firejail --build profile-file /opt/Notable/notable
then the following error occurs
/usr/bin/strace: Can't stat 'profile-file': Datei oder Verzeichnis nicht gefunden

<code>--build<b><i>=</i></b>profile-file</code>

**Reply by githlp** [reply]
thx, worked nearly.. the command then hangs, after aborting cannot fire again
```
firejail --build=profile-file /opt/Notable/notable
Error: the profile file already exists. Please use a different file name.
```

actually no notable.profile file was created (system wide). I tried to clean up with
`sudo firecfg`

but the error for profile building persists:
```
firejail --build=profile-file /opt/Notable/notable
Error: the profile file already exists. Please use a different file name.
```

#### As a workaround the following seems to be ok for me:
##### rm symlink
`rm /usr/bin/notable`

 ##### create desktop starter cmd
`firejail --noprofile --net=none /opt/Notable/notable %U`

##### add alias for terminal in .bash_aliases
`alias notable='firejail --noprofile --net=none /opt/Notable/notable &' `

**Reply by ghost** [reply]
> firejail --build=profile-file /opt/Notable/notable
Error: the profile file already exists. Please use a different file name.

> actually no notable.profile file was created (system wide)

That's not how this works. The file called `profile-file` will be placed in the directory from where you run that command, and if you don't move it out of the way firejail will keep complaining about it. The idea is that you carefully go over that file, using the profile template and test your app until you get a nicely tight sandbox you're happy with. To do that you can move it to `~/.config/firejail/notable.profile`. Mind the naming convention here: notable is the executable name, firejail expects to find a notable.profile so you can just start the sandbox by using `firejail /full/path/to/notable`. I'm simplifying here, but that's the basic idea: put as much of the sandboxing options in the profile and keep the start command as simple as possible.

> As a workaround the following seems to be ok for me:
[...]
firejail --noprofile --net=none /opt/Notable/notable %U

PLEASE DON'T USE `--noprofile`. It offers `zero protection` and is a debugging option. We should add that to the man page to make that crystal clear. If your app worked with the --noprofile option, it simply confirms firejail _can_ sandbox the app in question. So that's good news. How _restrictive_ the actual sandbox can be - for daily use of notable - will require thorough testing.

So please remove the desktop starter cmd and the alias, and post the content of your `profile-file` here, so we can start to understand what notable is doing exactly.

**Reply by rusty-snake** [reply]
FTR: Notable is an electron app.

**Reply by githlp** [reply]
ok, thx for making that clear.

the build option does not succeed (waited several  minutes) so a zero-byte sized profile-file is the result.
Sadly no error messages are provided by build, and strace of the command hangs right after the following output

```
execve("/usr/lib/firejail/fbuilder", ["/usr/lib/firejail/fbuilder", "--build=profile-file", "/opt/Notable/notable"], 0x558561a9ac10 /* 3 vars */) = 0
brk(NULL)                               = 0x55a94f341000
arch_prctl(0x3001 /* ARCH_??? */, 0x7fff60e736d0) = -1 EINVAL (the argument is invalid)
...
getuid()                                = 1000
access("profile-file", F_OK)            = -1 ENOENT (file or dir not found)
...
access("/usr/bin/strace", X_OK)         = 0
openat(AT_FDCWD, "/proc/sys/kernel/yama/ptrace_scope", O_RDONLY) = 4
newfstatat(4, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(4, "1\n", 1024)                    = 2
close(4)                                = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f249f9b5a10) = 8504
wait4(8504,  
```

So far.

**Comment by ghost** [comment, marked as answer]
I'll try to install the app and report back here when I can put together a firejail profile for it. Just to make sure, you're refering to this [notable](https://github.com/notable/notable) correct?

**Reply by githlp** [reply]
Yes, i am using latest = v1.8.4

Thank you.

**Reply by ghost** [reply]
Okay, I've been able to put something together. Before looking at/copying the below files to test your app, I want to point out that - as far as Firejail is concerned - there isn't anything really special about profiles for apps installed under /opt. As you've already hinted on, the only thing to remember is to add a launcher wrapper to your PATH, if your distro's package manager hasn't already taken care of that. In that context I advise to put the symlink you removed back in its original location, for ease of use.

I've only tested `notable` superficially, but all the basic functionality seems to be working fine here. Would be great if you would be able to test the app more thoroughly, as I'm assuming you're using it on a regular basis, or plan to. If things work out we can add a profile in master and ensure notable also gets into firecfg. But that's for later.

To test, copy the content of this to `~/.config/firejail/notable.profile`:

```
# Firejail profile for notable
# Description: The Markdown-based note-taking app that doesn't suck
# This file is overwritten after every install/update
# Persistent local customizations
include notable.local
# Persistent global definitions
include globals.local

# Note: On debian-based distributions the binary might be located in
# /opt/Notable/notable, and therefore not be in PATH.
# If that's the case you can start Notable with firejail via
# `firejail "/opt/Notable/notable"`.

mkdir ${HOME}/.config/Notable
mkdir ${HOME}/.notable
noblacklist ${HOME}/.config/Notable
noblacklist ${HOME}/.notable
whitelist ${HOME}/.config/Notable
whitelist ${HOME}/.notable
whitelist /opt/Notable

# block network access
net none
# no3d
nosound

# private-bin notable
?HAS_APPIMAGE: ignore private-dev
# private-lib

private-opt Notable

dbus-user filter
dbus-user.talk ca.desrt.dconf
ignore dbus-user none

# memory-deny-write-execute

# Redirect
include electron.profile

```

Notice these two lines

```
noblacklist ${HOME}/.config/Notable
noblacklist ${HOME}/.notable
```

The way Firejail works is to block all known locations for supported applications with a dedicated profile. That's the idea behind `disable-programs.inc`, which gets included by a lot of our profiles by default. So to mimic this behaviour, please add to/create a `~/.config/firejail/notable.local` with the below content:

```
# Firejail :: persistent local customizations for disable-programs.inc

# - - - - - - - - - - - - - - - - - - - - - - - - - - - -
#    keep until added upstream
# - - - - - - - - - - - - - - - - - - - - - - - - - - - -

# notable
blacklist ${HOME}/.notable
blacklist ${HOME}/.config/Notable

```
 
After taking care of the above; now is the time to start testing the sandbox.

```
$ firejail --ignore=quiet /usr/bin/notable
```

Happy testing!

**Reply by rusty-snake** [reply]
> whitelist /opt/Notable
> private-opt Notable

Logical conflict.

**Reply by ghost** [reply]
@rusty-snake Nice catch, forgot to take out `whitelist /opt/Notable`!

**Reply by ghost** [reply]
@githlp Here's the revised [notable.profile](https://gist.github.com/glitsj16/3ca8a83a4241751344c93bf09145c48f).

**Reply by githlp** [reply]
Thanks for creating a profile! The revised profile is missing the 
`# block network access`
`net none`
part though.

Notable starts but claims to be started for first time as it has no access (or is not aware) of all already existing notes (somenote.md) under
`~/.notable/notes/`

btw. i could not figure out why mousepad is not allowed to safe txt files under ~/.config/firejail - i went through all included .inc -files within the mousepad.profile file but did not spot a blacklist entry for this dir (?)

**Reply by rusty-snake** [reply]
> btw. i could not figure out why mousepad is not allowed to safe txt files under ~/.config/firejail - i went through all included .inc -files within the mousepad.profile file but did not spot a blacklist entry for this dir (?)

hardcoded `blacklist`

https://github.com/netblue30/firejail/blob/1000c238d5299a7fe3fc54a0d7001f8ce8979df8/src/firejail/fs.c#L825-L841

**Reply by ghost** [reply]
> Notable starts but claims to be started for first time as it has no access (or is not aware) of all already existing notes (somenote.md) under ~/.notable/notes/

Indeed, I missed that part of notable's behaviour in my earlier attempts at putting together a tight profile. What kind of voodoo this app (or its electron components) does escapes me (its source-code isn't publicly available AFAICT), but any attempt at whitelisting keeps showing that annoying behaviour. To fix it I had to drop back to blacklisting only, which is not something I'm happy about... For the record, here's the [newly adapted profile](https://gist.github.com/glitsj16/187162d642186e91d28831dd44371385).

It certainly isn't my intention to discourage you (or anyone else for that matter) from using `notable`. Yet, realizing its blacklisting sandbox is substandard, you might want to consider trying a note-taking app with IMHO much better [firejail support](https://github.com/netblue30/firejail/blob/master/etc/profile-m-z/marker.profile), like [marker](https://github.com/fabiocolacio/Marker) (which isn't electron-based).

**Reply by githlp** [reply]
Thx, seems to work now.

The marker app is similar, good app, though it seems to be less stable on a tiling wm and has no fuzzy search over all notes.
Since i have a large amount of markdown notes already i rather work with notable for the moment..

**Reply by ghost** [reply]
Thanks for confirming. I'll prepare a PR to add the notable profile soonish. Remember to check /etc/firejail after your OS package manager releases a new firejail package. If you see a /etc/firejail/notable.profile in there, you can remove ~/.config/firejail/notable.profile.
